src/lib/Inline.hs

```diff
@@ -0,0 +1,475 @@
+-- Copyright 2022 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+module Inline where
+
+import Data.List.NonEmpty qualified as NE
+
+import Data.Functor
+
+import Builder
+import Err
+import LabeledItems
+import Name
+import Occurrence hiding (Var)
+import Syntax
+
+-- === External API ===
+
+inlineBindings :: (EnvReader m) => SBlock n -> m n (SBlock n)
+inlineBindings blk = liftInlineM $ buildScopedAssumeNoDecls $ inline Stop blk
+{-# SCC inlineBindings #-}
+
+-- === Data Structure ===
+
+data SubstRange (o::S) where
+  DoneEx :: SExpr o -> SubstRange o
+  SuspEx :: SExpr i -> Subst InlineSubstVal i o -> SubstRange o
+
+instance Show (SubstRange n) where
+  show = \case
+    DoneEx e -> "finished " ++ show e
+    SuspEx e _ -> "unfinished " ++ show e
+
+instance SubstE Name SubstRange where
+  substE (scope, subst) = \case
+    DoneEx e -> DoneEx $ substE (scope, subst) e
+    SuspEx e s -> SuspEx e $ substE (scope, subst) s
+
+instance SinkableE SubstRange where
+  sinkingProofE rename = \case
+    DoneEx e   -> DoneEx $ sinkingProofE rename e
+    SuspEx e s -> SuspEx e $ sinkingProofE rename s
+
+type InlineSubstVal = SubstVal AtomNameC SubstRange
+
+newtype InlineM (i::S) (o::S) (a:: *) = InlineM
+  { runInlineM :: SubstReaderT InlineSubstVal (BuilderM SimpIR) i o a }
+  deriving ( Functor, Applicative, Monad, MonadFail, Fallible, ScopeReader
+           , EnvExtender, EnvReader, SubstReader InlineSubstVal, (Builder SimpIR)
+           , (ScopableBuilder SimpIR))
+
+liftInlineM :: (EnvReader m) => InlineM n n a -> m n a
+liftInlineM act = liftBuilder $ runSubstReaderT idSubst $ runInlineM act
+{-# INLINE liftInlineM #-}
+
+-- === Inliner ===
+
+data SizePreservationInfo =
+  -- Explicit noinline, or inlining doesn't preserve work
+    NoInline
+  -- In Secrets, this is either OnceSafe, or OnceUnsafe with whnfOrBot == True
+  | UsedOnce
+  -- In Secrets, this is either MultiSafe, or MultiUnsafe with whnfOrBot == True
+  | UsedMulti
+  deriving (Eq, Show)
+
+inlineDecls :: Emits o => Nest SDecl i i' -> InlineM i' o a -> InlineM i o a
+inlineDecls decls cont = case decls of
```

Comment:
BTW if you really care about the performance of this pass, this function should get refactored into a tail recursive one that doesn't call the continuation and only accumulates the substitution, and another INLINE wrapper around it that runs the helper and then calls the continuation. The latter way makes it clear to GHC that the continuation will be called exactly once. You can find this pattern e.g. [in Simplify](https://github.com/google-research/dex-lang/blob/main/src/lib/Simplify.hs#L174-L192).